src/context/index.ts

```diff
@@ -0,0 +1,232 @@
+		? (region.length === 4
+			? new Selection(...region)
+			: (() => {
+				const [byteOffset, byteLength] = region
+				const startColRaw = byteOffset % 16
+				const endColRaw = (byteOffset + byteLength) % 16
+				return new Selection(
+					Math.floor(byteOffset / 16),
+					10 + startColRaw + Math.floor(startColRaw / 2),
+					Math.floor((byteOffset + byteLength) / 16),
+					10 + endColRaw + Math.floor(endColRaw / 2),
+				)
+			})())
+		: (() => {
+			const line = doc.lineAt(region)
+			return new Selection(
+				line.range.start.line,
+				line.firstNonWhitespaceCharacterIndex,
+				line.range.end.line,
+				line.range.end.character,
+			)
+		})()
+	}
+
+export class Store {
+	static extensionPath: string | undefined
+	static globalState: Memento
+
+	@observable.shallow logs = [] as Log[]
+	@computed public get results() {
+		const runs = this.logs.map(log => log.runs).flat()
+		return runs.map(run => run.results).filter(run => run).flat() as Result[]
+	}
+	@computed public get distinctArtifactNames() {
+		const fileAndUris = this.logs.map(log => [...log._distinct.entries()]).flat()
+		return mapDistinct(fileAndUris)
+	}
+
+	constructor() {
+		intercept(this.logs, objChange => {
+			const change = objChange as unknown as IArrayWillSplice<Log>
+			change.added = change.added.filter(log => this.logs.every(existing => existing._uri !== log._uri))
+			return objChange
+		})
+	}
+}
+
+export async function activate(context: ExtensionContext) {
+	const disposables = context.subscriptions
+	Store.extensionPath = context.extensionPath
+	Store.globalState = context.globalState
+	disposables.push(commands.registerCommand('sarif.clearState', () => {
+		context.globalState.update('view', undefined)
+		commands.executeCommand('workbench.action.reloadWindow')
+	}))
+	const store = new Store()
+
+	// Boot
+	const uris = await workspace.findFiles('.sarif/**/*.sarif')
+	store.logs.push(...await loadLogs(uris))
+
+	// Basing
+	const urisNonSarif = await workspace.findFiles('**/*', '.sarif') // Ignore folders?
+	const fileAndUris = urisNonSarif.map(uri => [uri.path.split('/').pop(), uri.path])  as [string, string][]
+	const baser = new Baser(mapDistinct(fileAndUris), store)
+
+	// Panel
+	const panel = new Panel(context, baser, store)
+	if (uris.length) await panel.show()
+	disposables.push(commands.registerCommand('sarif.showPanel', () => panel.show()))
+
+	// Suggest In-Project Sarif Files
+	;(async () => {
+		const urisSarifInWorkspace = await workspace.findFiles('**/*.sarif', '.sarif/**/*.sarif')
+		const count = urisSarifInWorkspace.length
+		if (!count) return
+		if (await window.showInformationMessage(`Discovered ${count} SARIF logs in your workspace.`, 'View in SARIF Panel')) {
+			store.logs.push(...await loadLogs(urisSarifInWorkspace))
+			panel.show()
+		}
+	}) // Disabled while we evaluate the future of this feature.
+
+	// Diagnostics
+	const diagsAll = languages.createDiagnosticCollection('sarif')
+	const setDiags = (doc: TextDocument) => {
```

Comment:
>setDiags [](start = 7, length = 8)

I would prefer these functions be declared outside of this function so readers do not have to switch context from reading the activate function. The reader will lose the code flow of activate,